Meme/Meme/NetworkClient.swift

```diff
@@ -8,49 +8,64 @@
 
-protocol MemeClient {
-    func getMeme(from template: Template) -> Meme?
-}
-
-class NetworkClient: MemeClient {
-
-    fileprivate func getTemplates(_ data: Data?) -> [Template]? {
-        guard let data = data,
-            let jsonResponse = try? JSONSerialization.jsonObject(with: data, options: []) as? [String:String] else {
-                print("Error deserializing template data");
-                return nil
-        }
-        return jsonResponse.compactMap({ return Template.init(name: $0.0, templateURLString: $0.1)})
-    }
+class NetworkClient {
 
-    func requestTemplates(completion: @escaping ([Template]) -> Void)  {
+    func requestTemplates(completion: @escaping (Result<[Template], Error>) -> Void)  {
         let session = URLSession.shared
         let urlBuilder = MemeEndpointBuilder()
         let url = urlBuilder.endpointURL(.template)
         let request = URLRequest(url:url)
         let task: URLSessionDataTask = session.dataTask(with: request) { (data, response, error) -> Void in
-            guard let templates = self.getTemplates(data) else { return }
-            completion(templates)
+            guard let data = data else {
+                if let error = error {
+                    completion(.failure(error))
+                } else {
+                    if let responseError = (response as? HTTPURLResponse)?.apiError {
+                        completion(.failure(responseError))
+                    }
+                }
+                return
+            }
+
+            guard let jsonResponse = try? JSONSerialization.jsonObject(with: data, options: []) as? [String:String] else {
+                 let deserializationError = RetrievalError.deserializationFailure
+                    completion(.failure(deserializationError))
+                return
+            }
+            completion(.success(jsonResponse.compactMap({ return Template(name: $0.0, templateURLString: $0.1)})))
         }
         task.resume()
     }
+}
 
-    func requestMemes(completion: @escaping ([Meme]?) -> Void) {
-        requestTemplates { [weak self] templates in
-            completion(templates.compactMap({ return self?.getMeme(from: $0)}))
+extension HTTPURLResponse {
+    var apiError: NetworkError? {
+        switch statusCode {
+        case 100...199:
+            return .networkError
+        case 200...299:
+            return nil
+        case 300...399:
+            return .redirectionError
+        case 400...499:
+            return .clientError
+        case 500...599:
+            return .serverError
+        default:
+            return nil
         }
     }
+}
 
-    func getMeme(from template: Template) -> Meme? {
-        let urlBuilder = MemeEndpointBuilder()
-        guard let imageAlias = template.imageAlias,
-            let  imageURL = urlBuilder.imageURL(with: imageAlias) else {
-                print("Error constructing Meme model")
-                return nil
-        }
+enum NetworkError: Error {
```

Comment:
I wonder if we can consolidate all the errors here by adding .jsonDeserializationError and .unknownError? Since the first case is .networkError may have to rename the enum to something else. Maybe something like ClientError? Not sure

Reply:
@vhart `clientError` is also one of the cases so, maybe `MemeError`? I don't know if that makes it sound vague. `MemeClientError`, maybe?

Reply:
Yeah lets try that 👍 
The alternative is to nest the error types hierarchically but that's a bit much for what we need. Ill leave an example below (not the recommendation though)
```swift
enum NetworkError: Error {
    case ...
    case ...
    ....
}

enum MemeClientError: Error {
    case networkError(NetworkError)
    case deserializationError
    case unknownError
}
```